docs/proposals/kubegen.md

```diff
@@ -0,0 +1,214 @@
+# Kubernetes Code Generators
+
+Kubernetes provides a collection of code generators to assist with developing APIs.  These code generators generate
+useful libraries such as clients, openapi definitions, copying go structs, etc.
+
+The `kubegen` uses directory structure conventions to detect Kubernetes APIs definitions and run the standard
+set of Kubernetes code generators.
+
+There are 2 methods for running kubegen
+
+- Directly through the kubegen command line by downloading the binary and running at a the project root
+- Through Bazel by adding rules to `WORKSPACE` and `BUILD.gazel`
+
+## Kubernetes API conventions
+
+kubegen will run the full set of Kubernetes code generators for all APIs it finds in the project.
+kubegen looks for API groups defined under `pkg/apis/<group>/<version>` where *group* and *version* match the following
+patterns.
+
+- **group pattern**: ^[a-z]+$
+- **version pattern**: ^v\\d+(alpha\\d+|beta\\d+)*$
```

Comment:
hm, good point, I don't think that is ok.  Should probably be `^v\\d+(alpha\\d+|beta\\d+)?$`

Reply:
Why do we restrict the version pattern here?

Reply:
Same for the group pattern.

Reply:
I think it could be useful to restrict, since some generators assume things about how the version will be formatted, like [this one](https://github.com/kubernetes/code-generator/blob/master/cmd/client-gen/types/helpers.go#L52)

Reply:
@jennybuckley +1

I have found the code generators and apimachinery makes assumptions about the pattern and break unexpectedly and with obscure errors when they are not followed.

Reply:
It must be a valid Go identifier. But other than that? Restricting it to alpha/beta is definitely not necessary. If we require that anywhere, we should fix it.